File: apps/BarrelLock/BarrelLock_android/android/app/src/main/kotlin/com/hulk/bazaarAndroid/BackupP2pEventBridge.kt
package com.hulk.bazaarAndroid

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

/** Shared EventChannel sink for Nearby + BLE GATT backup sessions. */
object BackupP2pEventBridge : EventChannel.StreamHandler {
    private const val EVENT_CHANNEL = "com.barrellock/backup_p2p_events"

    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null

    fun register(messenger: BinaryMessenger) {
        EventChannel(messenger, EVENT_CHANNEL).setStreamHandler(this)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    fun emit(event: Map<String, Any?>) {
        mainHandler.post { eventSink?.success(event) }
    }
}

File: apps/ios/ios/Runner/BackupP2pEventBridge.swift
import Flutter

/// Shared EventChannel sink for Multipeer + BLE GATT backup sessions.
final class BackupP2pEventBridge: NSObject, FlutterStreamHandler {
  static let shared = BackupP2pEventBridge()
  static let channelName = "com.barrellock/backup_p2p_events"

  private var eventSink: FlutterEventSink?

  static func register(with registrar: FlutterPluginRegistrar) {
    let eventChannel = FlutterEventChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    eventChannel.setStreamHandler(shared)
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  func emit(_ event: [String: Any]) {
    DispatchQueue.main.async { [weak self] in
      self?.eventSink?(event)
    }
  }
}
